File: app/src/main/java/com/devidea/timeleft/ItemGenerate.java
package com.devidea.timeleft;

import java.time.Duration;
import java.time.LocalDate;
import java.time.LocalTime;
import java.time.format.TextStyle;
import java.time.temporal.ChronoUnit;
import java.util.Locale;

import static com.devidea.timeleft.MainActivity.appDatabase;

public class ItemGenerate implements InterfaceItem {

    @Override
    public AdapterItem timeItem() {
        AdapterItem adapterItem = new AdapterItem();

        LocalTime now = LocalTime.now();
        LocalTime endValue = LocalTime.parse("23:59:59");
        float secondsValue = Duration.between(LocalTime.of(0, 0), now).getSeconds();

        float TimePercent = (secondsValue / 86400) * 100;

        adapterItem.setSummery("오늘의 ");
        adapterItem.setPercentString(String.format(Locale.getDefault(), "%.1f", TimePercent));
        adapterItem.setLeftDay("남은시간: " + (LocalTime.ofSecondOfDay(Duration.between(now, endValue).getSeconds())));

        return adapterItem;
    }

    @Override
    public AdapterItem yearItem() {
        AdapterItem adapterItem = new AdapterItem();

        float YearPercent = ((float) LocalDate.now().getDayOfYear() / (float) LocalDate.now().lengthOfYear()) * 100;

        adapterItem.setSummery(LocalDate.now().getYear() + "년의 ");
        adapterItem.setPercentString(String.format(Locale.getDefault(), "%.1f", YearPercent));

        adapterItem.setLeftDay("남은일: " + (LocalDate.now().lengthOfYear() - LocalDate.now().getDayOfYear()) + "일");

        return adapterItem;
    }

    @Override
    public AdapterItem monthItem() {
        AdapterItem adapterItem = new AdapterItem();

        int monthOfDay = LocalDate.now().getDayOfMonth();
        int lengthOfMonth = LocalDate.now().lengthOfMonth();
        String summery = LocalDate.now().getMonth().getDisplayName(TextStyle.FULL, Locale.KOREAN);

        float MonthPercent = (float) monthOfDay / lengthOfMonth * 100;

        adapterItem.setSummery(summery + "의 ");
        adapterItem.setPercentString(String.format(Locale.getDefault(), "%.1f", MonthPercent));

        adapterItem.setLeftDay("남은일: " + (lengthOfMonth - monthOfDay) + "일");

        return adapterItem;
    }

    @Override
    public AdapterItem customTimeItem(EntityItemInfo itemInfo) {

        AdapterItem adapterItem = new AdapterItem();

        LocalTime startValue = LocalTime.parse(itemInfo.getStartValue()); //시작시간
        LocalTime endValue = LocalTime.parse(itemInfo.getEndValue()); // 종료시간
        LocalTime time = LocalTime.now(); //현재 시간

        adapterItem.setAutoUpdate(itemInfo.isAutoUpdate());
        adapterItem.setSummery(itemInfo.getSummery());
        adapterItem.setStartDay("설정시간: " + startValue);
        adapterItem.setEndDay("종료시간: " + endValue);
        adapterItem.setId(itemInfo.getId());

        if (time.isAfter(startValue) && time.isBefore(endValue)) {
            float range = Duration.between(startValue, endValue).getSeconds();
            float sendTime = Duration.between(startValue, time).getSeconds();

            adapterItem.setLeftDay("남은시간: " + (LocalTime.ofSecondOfDay(Duration.between(time, endValue).getSeconds())));

            float timePercent = (sendTime / range) * 100;

            adapterItem.setPercentString(String.format(Locale.getDefault(), "%.1f", timePercent));

        } else {
            adapterItem.setPercentString("100");
            adapterItem.setLeftDay("설정시간이 지나면 계산해 드릴께요");
        }

        return adapterItem;

    }

    @Override
    public AdapterItem customMonthItem(EntityItemInfo itemInfo) {

        AdapterItem adapterItem = new AdapterItem();

        LocalDate startDate = LocalDate.parse(itemInfo.getStartValue());
        LocalDate endDate = LocalDate.parse(itemInfo.getEndValue());
        LocalDate today = LocalDate.now();
        int id = itemInfo.getId();

        //설정일
        int setDay = (int) ChronoUnit.DAYS.between(startDate, endDate);

        //설정일에서 지난일
        int sendDay = (int) ChronoUnit.DAYS.between(startDate, today);

        //설정일까지 남은일
        int leftDay = (int) ChronoUnit.DAYS.between(today, endDate);
        int lengthOfMonth = LocalDate.now().lengthOfMonth();

        //종료일로 넘어가고 자동 업데이트 체크한 경우
        if (today.isAfter(endDate) && itemInfo.isAutoUpdate()) {
            //시작일: 종료일, 종료일: 종료일 + 설정일수로 UPDATE
            appDatabase.DatabaseDao().updateItem(String.valueOf(endDate), String.valueOf(endDate.plusDays(lengthOfMonth)), itemInfo.getId());

            itemInfo = appDatabase.DatabaseDao().getSelectItem(id);

            startDate = LocalDate.parse(itemInfo.getStartValue());
            endDate = LocalDate.parse(itemInfo.getEndValue());

            setDay = (int) ChronoUnit.DAYS.between(startDate, endDate);
            sendDay = (int) ChronoUnit.DAYS.between(startDate, today);
            leftDay = (int) ChronoUnit.DAYS.between(today, endDate);
        }

        float MonthPercent = (float) sendDay / setDay * 100;

        adapterItem.setStartDay("설정일: " + startDate);
        adapterItem.setEndDay("종료일: " + endDate);
        adapterItem.setLeftDay("남은일: D-" + leftDay);
        adapterItem.setAutoUpdate(itemInfo.isAutoUpdate());
        adapterItem.setSummery(itemInfo.getSummery());
        adapterItem.setId(itemInfo.getId());

        if (MonthPercent < 100) {
            adapterItem.setPercentString(String.format(Locale.getDefault(), "%.1f", MonthPercent));
        } else {
            adapterItem.setPercentString("100");
        }

        return adapterItem;
    }

}

File: app/src/main/java/com/devidea/timeleft/ItemGenerate.kt
package com.devidea.timeleft

import android.content.Context
import com.devidea.timeleft.calc.CustomTimeProgress
import com.devidea.timeleft.calc.TimeProgressCalculator
import com.devidea.timeleft.database.itemdata.ItemEntity
import com.devidea.timeleft.database.itemdata.ItemType
import com.devidea.timeleft.database.itemdata.RecurrenceMode
import dagger.hilt.android.qualifiers.ApplicationContext
import java.time.LocalDate
import java.time.LocalTime
import java.time.format.DateTimeFormatter
import java.time.format.TextStyle
import java.util.Locale
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class ItemGenerate @Inject constructor(
    @ApplicationContext private val context: Context,
) : InterfaceItem {

    override fun timeItem(): AdapterItem {
        val progress = TimeProgressCalculator.dayProgress(LocalTime.now())
        val leftFormatted = LocalTime.ofSecondOfDay(progress.durationLeft.seconds)
            .format(HEADER_TIME_FORMATTER)
        val leftText = context.getString(R.string.home_time_left, leftFormatted)

        return AdapterItem(
            title = context.getString(R.string.home_today_title),
            percent = roundPercent(progress.percentElapsed),
            leftString = leftText,
            widgetString = leftText.substring(0, leftText.length - 3),
        )
    }

    override fun yearItem(): AdapterItem {
        val today = LocalDate.now()
        val progress = TimeProgressCalculator.yearProgress(today)
        return AdapterItem(
            title = context.getString(R.string.home_year_title, today.year),
            percent = roundPercent(progress.percentElapsed),
            leftString = context.getString(R.string.home_days_left, progress.daysLeft),
        )
    }

    override fun monthItem(): AdapterItem {
        val today = LocalDate.now()
        val progress = TimeProgressCalculator.monthProgress(today)
        val monthName = today.month.getDisplayName(TextStyle.FULL, Locale.getDefault())
        return AdapterItem(
            title = context.getString(R.string.home_month_title, monthName),
            percent = roundPercent(progress.percentElapsed),
            leftString = context.getString(R.string.home_days_left, progress.daysLeft),
        )
    }

    override fun customTimeItem(itemEntity: ItemEntity): AdapterItem {
        val startTime = LocalTime.parse(itemEntity.startValue, STORAGE_TIME_FORMATTER)
        val endTime = LocalTime.parse(itemEntity.endValue, STORAGE_TIME_FORMATTER)

        val base = AdapterItem(
            id = itemEntity.id,
            title = itemEntity.title,
            startString = context.getString(R.string.card_time_start, startTime.toString()),
            endString = context.getString(R.string.card_time_end, endTime.toString()),
            updateInfo = context.getString(R.string.card_time_auto_start_hint),
            category = itemEntity.category,
            colorKey = itemEntity.colorKey,
            iconKey = itemEntity.iconKey,
            reminderText = reminderText(ItemType.Time, itemEntity.reminderOffsetDays),
        )

        return when (val result = TimeProgressCalculator.customTimeProgress(startTime, endTime, LocalTime.now())) {
            is CustomTimeProgress.Active -> {
                val leftFormatted = LocalTime.ofSecondOfDay(result.durationLeft.seconds)
                val leftText = context.getString(R.string.home_time_left, leftFormatted.toString())
                base.copy(
                    percent = roundPercent(result.percentElapsed),
                    leftString = leftText,
                    widgetString = leftText.substring(0, leftText.length - 3),
                    countdownText = leftFormatted.toString(),
                    dueText = context.getString(R.string.home_until_time, endTime.toString()),
                    remainingSortKey = result.durationLeft.seconds,
                )
            }
            CustomTimeProgress.Idle -> base.copy(
                percent = 100f,
                leftString = context.getString(R.string.card_time_idle_hint),
                widgetString = context.getString(R.string.card_time_widget_idle),
                countdownText = context.getString(R.string.home_waiting_countdown),
                dueText = context.getString(R.string.home_until_time, endTime.toString()),
            )
        }
    }

    override fun customMonthItem(itemEntity: ItemEntity): AdapterItem {
        val today = LocalDate.now()
        val startDate = LocalDate.parse(itemEntity.startValue, STORAGE_DATE_FORMATTER)
        val endDate = LocalDate.parse(itemEntity.endValue, STORAGE_DATE_FORMATTER)

        val progress = TimeProgressCalculator.customDateProgress(startDate, endDate, today)
        val displayPercent =
            if (progress.percentElapsed < 100f) roundPercent(progress.percentElapsed) else 100f

        val updateInfo = when (itemEntity.updateFlag) {
            RecurrenceMode.None ->
                context.getString(R.string.card_update_info_none)
            RecurrenceMode.Day ->
                context.getString(R.string.card_update_info_day, itemEntity.updateRate)
            RecurrenceMode.Month ->
                context.getString(R.string.card_update_info_month, itemEntity.updateRate)
            RecurrenceMode.TimeRange -> ""
        }
        val recurrenceText = when (itemEntity.updateFlag) {
            RecurrenceMode.Day -> context.getString(R.string.card_recurrence_day, itemEntity.updateRate)
            RecurrenceMode.Month -> context.getString(R.string.card_recurrence_month, itemEntity.updateRate)
            RecurrenceMode.None, RecurrenceMode.TimeRange -> ""
        }

        val countdownText = ddayText(progress.daysLeft)

        return AdapterItem(
            id = itemEntity.id,
            title = itemEntity.title,
            startString = context.getString(R.string.card_date_start, startDate.toString()),
            endString = context.getString(R.string.card_date_end, endDate.toString()),
            leftString = context.getString(R.string.card_dday_value, countdownText),
            percent = displayPercent,
            updateInfo = updateInfo,
            countdownText = countdownText,
            dueText = context.getString(R.string.home_until_date, endDate.toString()),
            recurrenceText = recurrenceText,
            category = itemEntity.category,
            colorKey = itemEntity.colorKey,
            iconKey = itemEntity.iconKey,
            reminderText = reminderText(ItemType.Date, itemEntity.reminderOffsetDays),
            remainingSortKey = if (progress.daysLeft >= 0) progress.daysLeft.toLong() * SECONDS_PER_DAY else Long.MAX_VALUE,
            isExpired = progress.daysLeft < 0,
        )
    }

    private fun reminderText(type: ItemType, offset: Int): String =
        if (offset == ItemVisuals.REMINDER_DISABLED) {
            ""
        } else {
            context.getString(ItemVisuals.reminderNameRes(type, offset))
        }

    private fun roundPercent(raw: Float): Float =
        String.format(Locale.getDefault(), "%.1f", raw).toFloat()

    private fun ddayText(daysLeft: Int): String =
        when {
            daysLeft > 0 -> context.getString(R.string.home_dday_before, daysLeft)
            daysLeft == 0 -> context.getString(R.string.home_dday_today)
            else -> context.getString(R.string.home_dday_after, -daysLeft)
        }

    companion object {
        private const val SECONDS_PER_DAY = 86_400L
        private val HEADER_TIME_FORMATTER = DateTimeFormatter.ofPattern("H:mm:ss")
        private val STORAGE_TIME_FORMATTER = DateTimeFormatter.ofPattern("H:m")
        private val STORAGE_DATE_FORMATTER = DateTimeFormatter.ofPattern("yyyy-M-d")
    }
}
